// @stdlib/array/fixed-endian-factory/test/test.sort.js
/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var hasOwnProp = require( '@stdlib/assert/has-own-property' );
var hasSameValues = require( '@stdlib/array/base/assert/has-same-values' );
var isFunction = require( '@stdlib/assert/is-function' );
var factory = require( './../lib' );


// FUNCTIONS //

/**
* Comparison function.
*
* @private
* @param {number} a - first value for comparison
* @param {number} b - second value for comparison
* @returns {number} comparison result
*/
function compareFcn( a, b ) {
	if ( a > b ) {
		return 1;
	}
	if ( a < b) {
		return -1;
	}
	return 0;
}


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof factory, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns a function', function test( t ) {
	var ctor = factory( 'float64' );
	t.strictEqual( isFunction( ctor ), true, 'returns expected value' );
	t.end();
});

tape( 'attached to the prototype of the returned function is an `sort` method', function test( t ) {
	var ctor = factory( 'float64' );
	t.strictEqual( hasOwnProp( ctor.prototype, 'sort' ), true, 'returns expected value' );
	t.strictEqual( isFunction( ctor.prototype.sort ), true, 'returns expected value' );
	t.end();
});

tape( 'the method throws an error if invoked with a `this` context which is not a typed array instance', function test( t ) {
	var values;
	var ctor;
	var arr;
	var i;

	ctor = factory( 'float64' );
	arr = new ctor( 'little-endian', [ 1.0, 2.0, 3.0, 4.0, 5.0 ] );

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		{},
		[],
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			return arr.sort.call( value, compareFcn );
		};
	}
});

tape( 'the method throws an error if provided a first argument which is not a function', function test( t ) {
	var values;
	var ctor;
	var arr;
	var i;

	ctor = factory( 'float64' );
	arr = new ctor( 'little-endian', [ 1.0, 2.0, 3.0, 4.0 ] );

	values = [
		'5',
		3.14,
		NaN,
		true,
		false,
		null,
		void 0,
		{},
		[]
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			return arr.sort( value );
		};
	}
});

tape( 'the method returns an empty array if operating on an empty array', function test( t ) {
	var ctor;
	var out;
	var arr;

	ctor = factory( 'float64' );
	arr = new ctor( 'little-endian' );
	out = arr.sort( compareFcn );

	t.strictEqual( out.buffer, arr.buffer, 'returns expected value' );
	t.strictEqual( out.length, 0, 'returns expected value' );
	t.end();
});

tape( 'the method sorts elements of a typed array in-place', function test( t ) {
	var expected;
	var ctor;
	var arr;
	var out;

	ctor = factory( 'float64' );
	arr = new ctor( 'little-endian', [ 5.0, 4.0, 1.0, 2.0, 3.0 ] );
	expected = new ctor( 'little-endian', [ 1.0, 2.0, 3.0, 4.0, 5.0 ] );
	out = arr.sort( compareFcn );

	t.strictEqual( out instanceof ctor, true, 'returns expected value' );
	t.strictEqual( out, arr, 'returns expected value' );
	t.strictEqual( arr.length, expected.length, 'returns expected value' );
	t.strictEqual( hasSameValues( out, expected ), true, 'returns expected value' );

	t.end();
});

tape( 'the method sorts elements of a typed array without passing a comparison function', function test( t ) {
	var expected;
	var ctor;
	var arr;
	var out;

	ctor = factory( 'float64' );
	arr = new ctor( 'little-endian', [ 5.0, 4.0, 1.0, 2.0, 3.0 ] );
	expected = new ctor( 'little-endian', [ 1.0, 2.0, 3.0, 4.0, 5.0 ] );
	out = arr.sort();
	t.strictEqual( out.length, expected.length, 'returns expected value' );
	t.strictEqual( hasSameValues( out, expected ), true, 'returns expected value' );
	t.end();
});

tape( 'the method does not change the array length', function test( t ) {
	var ctor;
	var arr;
	var out;

	ctor = factory( 'float64' );
	arr = new ctor( 'little-endian', [ 5.0, 4.0, 1.0, 2.0, 3.0 ] );
	out = arr.sort();

	t.strictEqual( out.length, 5, 'returns expected value' );
	t.end();
});
